tools/perfscale-audit/metric-client/metric-client.go

```diff
@@ -113,6 +116,50 @@ type metric struct {
 	timestamp time.Time
 }
 
+func calculateOffset(endTime time.Time, waitTime time.Duration, scrapeInterval time.Duration) int {
+	// Always ensure we get at least to last Prometheus scrape by adding
+	// the PrometheusScrapeInterval to the offset
+	testEnd := endTime.Add(waitTime)
+	lookBack := int(time.Now().Sub(testEnd).Seconds()) + int(scrapeInterval.Seconds())
+
+	if lookBack < 1 {
+		lookBack = int(scrapeInterval.Seconds())
+	}
+	return lookBack
+}
+
+func calculateRangeVector(scrapeInterval time.Duration, testDuration time.Duration) time.Duration {
+	var rv time.Duration
+
+	// We're going to use a range vector that's 10x as long as the scrape interval.
+	// E.g. 30s scrapeInterval means [5m] range vector.  This will give the most
+	// reasonable results from interpolation.
+	rv = time.Duration(10 * scrapeInterval)
+
+	// When the range vector is shorter than the testDuration, use the testDuration
+	// for the range vector to improve test accuracy.
+	if rv.Seconds() <= testDuration.Seconds() {
+		log.Printf("rv %v duration %v\n", rv, testDuration)
+		return testDuration
+	} else {
+		// When the testDuration is less than the rangeVector, there's risk
+		// for the current test data will get mixed into future test data.
+		// So Sleep() until testDuration is equal to the range vector.  This
+		// means with the default scape interval, tests will be 5 minutes
+		// in length
+		waitTime := rv.Seconds() - testDuration.Seconds()
+		log.Printf("Sleeping for %vs so range vector is at least greater than or equal to the testDuration", waitTime)
```

Comment:
You have offset logic to move the scrape interval back to the experiment end time to not collect idle time after the experiment ends. Otherwise, without the offset, it will collect from the start of the test until the moment when the audit tool is running....

Will sleeping 5 minutes after the test ends affect anything since we have the offset logic?

Reply:
The affect offset has is it will make the test result more accurate.  For example with the 5m sleep, `testEnd` is  `endTime` plus the `waitTime`.  So in this case, the timestamp it returns will be `5m` after `startTime`.  I add the time between `endTime` and when this query was run to the offset in case the code was slow and add `PrometheusScrapeInterval` .  The result is the query is run near the end of the test to guarantee the query gets data.

```go
	testEnd := endTime.Add(waitTime)
	lookBack := int(time.Now().Sub(testEnd).Seconds()) + int(scrapeInterval.Seconds())
```

Reply:
Make sense, thanks!